### PR TITLE
AnimSplineIK node added to inverse kinematics

### PR DESCRIPTION
**Pull request by amantley**
This pr is an opitimization that is being done for the Quest build.
We are adding an AnimSplineIK node to the animation graph and removing the old AnimInverseKinematicsNode

Unit Tests are currently being developed but this code is ready for dev review.

### CODE REVIEW

**Comment by hifi-gustavo** [comment]
Windows build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/16970/HighFidelity-Beta-PR14817-61b019d.exe

**Comment by hifi-gustavo** [comment]
Windows client-only build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/16970/HighFidelity-Beta-Interface-PR14817-61b019d.exe

**Comment by hifi-gustavo** [comment]
Windows build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/16982/HighFidelity-Beta-PR14817-25c5a2f.exe

**Comment by hifi-gustavo** [comment]
Windows client-only build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/16982/HighFidelity-Beta-Interface-PR14817-25c5a2f.exe

**Comment by hifi-gustavo** [comment]
Android build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dandroid/16984/HighFidelity-Beta-PR14817-dd99f93.apk

**Comment by hifi-gustavo** [comment]
Windows build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/16984/HighFidelity-Beta-PR14817-dd99f93.exe

**Comment by hifi-gustavo** [comment]
Windows client-only build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/16984/HighFidelity-Beta-Interface-PR14817-dd99f93.exe

**Comment by hifi-gustavo** [comment]
Android build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dandroid/16985/HighFidelity-Beta-PR14817-a2ef7ed.apk

File: interface/resources/avatar/avatar-animation_withSplineIKNode.json [hunk 1]
```diff
@@ -0,0 +1,2247 @@
...
+                                                                                    "tipRotationVar": "headRotation",
+                                                                                    "alphaVar": "splineIKAlpha",
+                                                                                    "enabledVar": "splineIKEnabled",
+                                                                                    "tipTargetFlexCoefficients": [ 1.0, 1.0, 1.0, 1.0, 1.0 ],
+                                                                                    "midTargetFlexCoefficients": [ 1.0, 1.0, 1.0 ]
+                                                                                },
+                                                                                "children": [
+                                                                                    {
+                                                                                        "id": "ikOverlay",
+                                                                                        "type": "overlay",
```
Comment (hyperlogic) [inline interface/resources/avatar/avatar-animation_withSplineIKNode.json]: You can get rid of the "ikOverlay" node.

File: interface/resources/avatar/avatar-animation_withSplineIKNode.json [hunk 2]
```diff
@@ -0,0 +1,2247 @@
...
+                        "data": {
+                            "alpha": 1.0,
+                            "enabled": false,
+                            "interpDuration": 15,
+                            "baseJointName": "RightUpLeg",
+                            "midJointName": "RightLeg",
+                            "tipJointName": "RightFoot",
+                            "midHingeAxis": [ -1, 0, 0 ],
+                            "alphaVar": "rightFootIKAlpha",
+                            "enabledVar": "rightFootIKEnabled",
```
Comment (hyperlogic) [inline interface/resources/avatar/avatar-animation_withSplineIKNode.json]: In Rig::computeMotionAnimationState() there's a section of code that is used to enable and disable IK.  This is triggered from the Developer > Enable Inverse Kinematics menu.  We still want that menu to work, admittatly it probably doesn't work completely even on master.  Can you make sure that this code sets animVars such as rightFootIKEnabled to properly turn off IK for all the TwoBoneIK PoleVectorConstaint and SplineIK nodes.

File: libraries/animation/src/AnimSplineIK.h [hunk 3]
```diff
@@ -0,0 +1,104 @@
...
+    struct SplineJointInfo {
+        int jointIndex;       // joint in the skeleton that this information pertains to.
+        float ratio;          // percentage (0..1) along the spline for this joint.
+        AnimPose offsetPose;  // local offset from the spline to the joint.
+    };
+
+    void solveTargetWithSpline(const AnimContext& context, int base, const IKTarget& target, const AnimPoseVec& absolutePoses, bool debug, AnimChain& chainInfoOut) const;
+    void computeAndCacheSplineJointInfosForIKTarget(const AnimContext& context, int base, const IKTarget& target) const;
+    const std::vector<SplineJointInfo>* findOrCreateSplineJointInfo(const AnimContext& context, int base, const IKTarget& target) const;
+    mutable std::map<int, std::vector<SplineJointInfo>> _splineJointInfoMap;
```
Comment (hyperlogic) [inline libraries/animation/src/AnimSplineIK.h]: I'm not sure you need this map anymore, just the vector.  This is only necessary in the AnimInverseKinematicsNode where there is more than one spline.

**Comment by hifi-gustavo** [comment]
Android build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dandroid/17037/HighFidelity-Beta-PR14817-af38a40.apk

**Comment by hyperlogic** [comment]
https://highfidelity.manuscript.com/f/cases/20599/IK-for-Quest-Spline-IK-Node
https://highfidelity.manuscript.com/f/cases/20600/IK-for-Quest-Create-Quest-Animation-json
https://highfidelity.manuscript.com/f/cases/20604/IK-for-Quest-Refine-Elbow-Rotations

**Comment by hifi-gustavo** [comment]
Android build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dandroid/17345/HighFidelity-Beta-PR14817-3609392.apk

**Comment by hifi-gustavo** [comment]
Android build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dandroid/17364/HighFidelity-Beta-PR14817-d78f253.apk

**Comment by hifi-gustavo** [comment]
Windows build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/17364/HighFidelity-Beta-PR14817-d78f253.exe

**Comment by hifi-gustavo** [comment]
Windows client-only build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/17364/HighFidelity-Beta-Interface-PR14817-d78f253.exe

**Comment by hifi-gustavo** [comment]
Android build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dandroid/17373/HighFidelity-Beta-PR14817-0982c37.apk

**Comment by hifi-gustavo** [comment]
Windows build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/17373/HighFidelity-Beta-PR14817-0982c37.exe

**Comment by hifi-gustavo** [comment]
Windows client-only build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/17373/HighFidelity-Beta-Interface-PR14817-0982c37.exe

**Comment by hifi-gustavo** [comment]
Android build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dandroid/17388/HighFidelity-Beta-PR14817-f2301e7.apk

**Comment by hifi-gustavo** [comment]
Windows build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/17388/HighFidelity-Beta-PR14817-f2301e7.exe

**Comment by hifi-gustavo** [comment]
Windows client-only build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/17388/HighFidelity-Beta-Interface-PR14817-f2301e7.exe

**Comment by hifi-gustavo** [comment]
Android build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dandroid/17442/HighFidelity-Beta-PR14817-951380d.apk

**Comment by hifi-gustavo** [comment]
Android build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dandroid/17443/HighFidelity-Beta-PR14817-7639eac.apk

**Comment by hifi-gustavo** [comment]
Windows build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/17443/HighFidelity-Beta-PR14817-7639eac.exe

**Comment by hifi-gustavo** [comment]
Windows client-only build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/17443/HighFidelity-Beta-Interface-PR14817-7639eac.exe

**Comment by hifi-gustavo** [comment]
Android build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dandroid/17470/HighFidelity-Beta-PR14817-27bfe2f.apk

**Comment by hifi-gustavo** [comment]
Windows build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/17470/HighFidelity-Beta-PR14817-27bfe2f.exe

**Comment by hifi-gustavo** [comment]
Windows client-only build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/17470/HighFidelity-Beta-Interface-PR14817-27bfe2f.exe

**Comment by hifi-gustavo** [comment]
Android build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dandroid/17474/HighFidelity-Beta-PR14817-f8a74ef.apk

**Comment by hifi-gustavo** [comment]
Windows build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/17474/HighFidelity-Beta-PR14817-f8a74ef.exe

**Comment by hifi-gustavo** [comment]
Windows client-only build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/17474/HighFidelity-Beta-Interface-PR14817-f8a74ef.exe

**Comment by hifi-gustavo** [comment]
Android build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dandroid/17476/HighFidelity-Beta-PR14817-b6bc467.apk

**Comment by hifi-gustavo** [comment]
Windows build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/17476/HighFidelity-Beta-PR14817-b6bc467.exe

**Comment by hifi-gustavo** [comment]
Windows client-only build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/17476/HighFidelity-Beta-Interface-PR14817-b6bc467.exe

**Comment by hifi-gustavo** [comment]
Android build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dandroid/17477/HighFidelity-Beta-PR14817-33fd64c.apk

**Comment by hifi-gustavo** [comment]
Windows build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/17477/HighFidelity-Beta-PR14817-33fd64c.exe

**Comment by hifi-gustavo** [comment]
Windows client-only build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/17477/HighFidelity-Beta-Interface-PR14817-33fd64c.exe

File: libraries/animation/src/AnimPoleVectorConstraint.h [hunk 4]
```diff
@@ -25,6 +25,8 @@ class AnimPoleVectorConstraint : public AnimNode {
                              const QString& enabledVar, const QString& poleVectorVar);
     virtual ~AnimPoleVectorConstraint() override;
 
+    float findThetaNewWay(const glm::vec3& hand, const glm::vec3& shoulder, bool left) const;
```
Comment (hyperlogic) [inline libraries/animation/src/AnimPoleVectorConstraint.h]: remove

File: libraries/animation/src/AnimPoleVectorConstraint.h [hunk 5]
```diff
@@ -64,6 +66,8 @@ class AnimPoleVectorConstraint : public AnimNode {
     float _interpAlphaVel { 0.0f };
     float _interpAlpha { 0.0f };
 
+    float _lastTheta { 0.0f };
```
Comment (hyperlogic) [inline libraries/animation/src/AnimPoleVectorConstraint.h]: remove

File: libraries/animation/src/AnimPoleVectorConstraint.cpp [hunk 6]
```diff
@@ -117,7 +117,7 @@ const AnimPoseVec& AnimPoleVectorConstraint::evaluate(const AnimVariantMap& anim
     if (axisLength > MIN_LENGTH && refVectorLength > MIN_LENGTH && sideVectorLength > MIN_LENGTH &&
         refVectorProjLength > MIN_LENGTH && poleVectorProjLength > MIN_LENGTH) {
 
-        float dot = glm::clamp(glm::dot(refVectorProj / refVectorProjLength, poleVectorProj / poleVectorProjLength), 0.0f, 1.0f);
+        float dot = glm::clamp(glm::dot(refVectorProj / refVectorProjLength, poleVectorProj / poleVectorProjLength), -1.0f, 1.0f);
```
Comment (hyperlogic) [inline libraries/animation/src/AnimPoleVectorConstraint.cpp]: good catch

File: libraries/avatars-renderer/src/avatars-renderer/Avatar.h [hunk 7]
```diff
@@ -25,6 +25,7 @@
 #include <render/Scene.h>
 #include <graphics-scripting/Forward.h>
 #include <GLMHelpers.h>
+#include <CubicHermiteSpline.h>
```
Comment (hyperlogic) [inline libraries/avatars-renderer/src/avatars-renderer/Avatar.h]: Move this to Avatar.cpp file.  It's not necessary in the header.

File: libraries/shared/src/AvatarConstants.h [hunk 8]
```diff
@@ -20,6 +20,7 @@ const float DEFAULT_AVATAR_EYE_TO_TOP_OF_HEAD = 0.11f; // meters
 const float DEFAULT_AVATAR_NECK_TO_TOP_OF_HEAD = 0.185f; // meters
 const float DEFAULT_AVATAR_NECK_HEIGHT = DEFAULT_AVATAR_HEIGHT - DEFAULT_AVATAR_NECK_TO_TOP_OF_HEAD;
 const float DEFAULT_AVATAR_EYE_HEIGHT = DEFAULT_AVATAR_HEIGHT - DEFAULT_AVATAR_EYE_TO_TOP_OF_HEAD;
+const float DEFAULT_SPINE2_SPLINE_PROPORTION = 0.75f;
```
Comment (hyperlogic) [inline libraries/shared/src/AvatarConstants.h]: Can you double check this number against the "The Measure of Man and Woman: Human Factors in Design" book

File: libraries/shared/src/CubicHermiteSpline.h [hunk 9]
```diff
@@ -81,6 +81,27 @@ class CubicHermiteSplineFunctorWithArcLength : public CubicHermiteSplineFunctor
...
 
+    CubicHermiteSplineFunctorWithArcLength(const glm::quat& tipRot, const glm::vec3& tipTrans, const glm::quat& baseRot, const glm::vec3& baseTrans, float baseGain = 1.0f, float tipGain = 1.0f) : CubicHermiteSplineFunctor() {
+
+        float linearDistance = glm::length(baseTrans - tipTrans);
+        _p0 = baseTrans;
+        _m0 = baseGain * linearDistance * (baseRot * Vectors::UNIT_Y);
+        _p1 = tipTrans;
+        _m1 = tipGain * linearDistance * (tipRot * Vectors::UNIT_Y);
+
+        // initialize _values with the accumulated arcLength along the spline.
```
Comment (hyperlogic) [inline libraries/shared/src/CubicHermiteSpline.h]: Can you move this block that computes arcLengthValues into a protected function. that can be called in both CubicHermiteSplineFunctorWithArcLength constructors.

File: CMakeLists.txt [hunk 10]
```diff
@@ -52,6 +52,7 @@ else()
   set(MOBILE 0)
 endif()
 
+set(HIFI_USE_Q_OS_ANDROID_OPTION OFF)
```
Comment (hyperlogic) [inline CMakeLists.txt]: I want this preprocessor define to be named HIFI_USE_OPTIMIZED_IK  not HIFI_USE_Q_OS_ANDROID

**Comment by hyperlogic** [comment]
Also, please fix the Linux and Mac build warnings.

**Comment by hifi-gustavo** [comment]
Android build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dandroid/17496/HighFidelity-Beta-PR14817-c9db426.apk

**Comment by hifi-gustavo** [comment]
Windows build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/17496/HighFidelity-Beta-PR14817-c9db426.exe

**Comment by hifi-gustavo** [comment]
Windows client-only build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/17496/HighFidelity-Beta-Interface-PR14817-c9db426.exe

**Comment by hifi-gustavo** [comment]
Android build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dandroid/17498/HighFidelity-Beta-PR14817-5cf8a96.apk

**Comment by hifi-gustavo** [comment]
Windows build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/17498/HighFidelity-Beta-PR14817-5cf8a96.exe

**Comment by hifi-gustavo** [comment]
Windows client-only build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/17498/HighFidelity-Beta-Interface-PR14817-5cf8a96.exe

**Comment by hifi-gustavo** [comment]
Android build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dandroid/17500/HighFidelity-Beta-PR14817-3bf5c44.apk

**Comment by hifi-gustavo** [comment]
Mac OS X build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/17500/HighFidelity-Beta-PR14817-3bf5c44.dmg

**Comment by hifi-gustavo** [comment]
Mac OS X client-only build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/17500/HighFidelity-Beta-Interface-PR14817-3bf5c44.dmg

**Comment by hifi-gustavo** [comment]
Windows build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/17500/HighFidelity-Beta-PR14817-3bf5c44.exe

**Comment by hifi-gustavo** [comment]
Windows client-only build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/17500/HighFidelity-Beta-Interface-PR14817-3bf5c44.exe

**Pull request merged**